feat(gen): support custom time format

abemedia opened the pull request:
This PR adds support for a new extension `x-ogen-time-format` which sets a format for encoding and decoding time values.
Also fixes some indentation on templates (mix of tabs and spaces).

Since I'm not that familiar with the codebase, let me know if I should do some things differently.

Closes #1063

tdakkota commented on json/time.go:
```diff
@@ -7,71 +7,38 @@ import (
-	s.ByteStr(b)
+// EncodeTime encodes date, time & date-time to json.
+func EncodeTime(s *jx.Encoder, v time.Time, layout string) {
+	s.Str(v.Format(layout))
```
Change of function signature is a breaking change and these functions are public and released under v1 version.
I think it is better to keep existing function and add a new one  with layout argument. Old function may call a new one with `DateLayout` argument.

Old version with stack allocated buffer is much faster on my PC:
```
$ git checkout main
$ go test -count=15 -benchmem -bench='BenchmarkEncodeTime' -run='^$' ./json > old.txt
$ gh pr checkout 1334
$ go test -count=15 -benchmem -bench='BenchmarkEncodeTime' -run='^$' ./json > new.txt
$ benchstat old.txt new.txt
goos: linux
goarch: amd64
pkg: github.com/ogen-go/ogen/json
cpu: AMD Ryzen 9 5950X 16-Core Processor            
              │   old.txt   │                new.txt                │
              │   sec/op    │    sec/op     vs base                 │
EncodeTime-32   50.37n ± 1%   124.30n ± 4%  +146.77% (p=0.000 n=15)

              │  old.txt  │           new.txt            │
              │   B/op    │    B/op     vs base          │
EncodeTime-32   0.00 ± 0%   32.00 ± 0%  ? (p=0.000 n=15)

              │  old.txt   │           new.txt            │
              │ allocs/op  │ allocs/op   vs base          │
EncodeTime-32   0.000 ± 0%   1.000 ± 0%  ? (p=0.000 n=15)
$ go version
go version go1.23.2 linux/amd64
```

So, from performance standpoint, it's better to do something like:
```go
func EncodeTimeLayout(e *jx.Encoder, v time.Time, layout string) {
	const stackThreshold = 64

	var buf []byte
	if len(layout) > stackThreshold {
		buf = make([]byte, len(layout))
	} else {
		// Allocate buf on stack, if we can.
		buf = make([]byte, stackThreshold)
	}

	buf = v.AppendFormat(buf[:0], layout)
	e.ByteStr(buf)
}

func EncodeDateTime(e *jx.Encoder, v time.Time) {
	EncodeTimeLayout(e, v, time.RFC3339)
}
```

tdakkota reviewed (changes requested):
Thank you for your contribution!

Please move [generated code changes to a separate commit](https://github.com/ogen-go/ogen/blob/main/CONTRIBUTING.md#committing-generated-code).

Also, it would be nice to have a encode-decode-encode test for this feature, [like one of these](https://github.com/ogen-go/ogen/blob/main/internal/integration/json_test.go). 

Everything else LGTM.

abemedia commented:
@tdakkota Weirdly when testing on my Intel MacBook I can reproduce similar results as you on the benchmarks. On my M1 `Format` was actually faster (however the function you wrote in your comments beat it by a little).

Since we're leaving the `json` package functions I had removed, I changed `TimeFormat` to only return custom formats so the output of the template no longer changes if no custom format was set.

Also, I had forgotten parameters on the last PR so I added support for those too.

Let me know if there's anything else.
